arroyo-worker/src/connectors/two_phase_committer.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+#[process_fn(in_k = K, in_t = T)]
+impl<K: Key, T: Data + Sync, TPC: TwoPhaseCommitter<K, T>> TwoPhaseCommitterOperator<K, T, TPC> {
+    pub(crate) fn new(committer: TPC) -> Self {
+        Self {
+            committer,
+            pre_commits: Vec::new(),
+            phantom: PhantomData,
+        }
+    }
+
+    fn name(&self) -> String {
+        self.committer.name()
+    }
+
+    fn tables(&self) -> Vec<arroyo_rpc::grpc::TableDescriptor> {
+        vec![
+            arroyo_state::global_table("r", "recovery data"),
+            TableDescriptor {
+                name: "p".into(),
+                description: "pre-commit data".into(),
+                table_type: TableType::Global as i32,
+                delete_behavior: TableDeleteBehavior::None as i32,
+                write_behavior: TableWriteBehavior::CommitWrites as i32,
+                retention_micros: 0,
+            },
+        ]
+    }
+
+    async fn on_start(&mut self, ctx: &mut Context<(), ()>) {
+        let mut tracking_key_state: GlobalKeyedState<
+            usize,
+            <TPC as TwoPhaseCommitter<K, T>>::DataRecovery,
+            _,
+        > = ctx.state.get_global_keyed_state('r').await;
+        // take the max of all values
+        let state_vec = tracking_key_state
+            .get_all()
+            .into_iter()
+            .map(|state| state.clone())
+            .collect();
+        self.committer
+            .init(&ctx.task_info, state_vec)
+            .await
+            .expect("committer initialized");
+
+        // subtask 0 is responsible for finishing commits if we were interrupted mid commit.
+        if ctx.task_info.task_index == 0 {
+            let mut pre_commit_state: GlobalKeyedState<
+                String,
+                <TPC as TwoPhaseCommitter<K, T>>::PreCommit,
+                _,
+            > = ctx.state.get_global_keyed_state('p').await;
+            self.pre_commits = pre_commit_state
+                .get_all()
+                .into_iter()
+                .map(|state| state.clone())
+                .collect();
+        }
+    }
+
+    async fn process_element(&mut self, record: &Record<K, T>, _ctx: &mut Context<(), ()>) {
+        self.committer
+            .insert_record(record)
+            .await
+            .expect("record inserted");
+    }
+
+    async fn on_close(&mut self, ctx: &mut crate::engine::Context<(), ()>) {
+        info!("waiting for commit message");
+        if let Some(ControlMessage::Commit { epoch }) = ctx.control_rx.recv().await {
+            self.handle_commit(epoch, ctx).await;
+        } else {
+            warn!("no commit message received, not committing")
+        }
+    }
+
+    async fn handle_checkpoint(
+        &mut self,
+        checkpoint_barrier: &arroyo_types::CheckpointBarrier,
+        ctx: &mut crate::engine::Context<(), ()>,
+    ) {
+        info!("received checkpoint barrier {:?}", checkpoint_barrier);
```

Comment:
should this be an info log? I think the underlying infrastructure already logs this.

Reply:
There's a number of logging statements that I intend to remove before merging but were helpful for debugging.